src/docfx/restore/CredentialProvider.cs

```diff
@@ -4,36 +4,92 @@
 using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Linq;
 using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace Microsoft.Docs.Build
 {
     internal class CredentialProvider
     {
-        private readonly IReadOnlyDictionary<string, HttpConfig> _credentials;
-        private readonly ConcurrentDictionary<string, Dictionary<string, string>> _credentialCache = new();
+        private static readonly Dictionary<string, HttpConfig> s_refreshedCredentials = new();
+        private static readonly SemaphoreSlim s_credentialRefreshSemaphore = new(1);
 
-        public CredentialProvider(IReadOnlyDictionary<string, HttpConfig> credentials)
+        private static readonly ConcurrentDictionary<string, Dictionary<string, string>> s_credentialCache = new();
+
+        private readonly IReadOnlyDictionary<string, HttpConfig> _configCredentials;
+        private readonly Func<string, Task<Dictionary<string, HttpConfig>>>? _getCredential;
+
+        public bool SupportRefreshToken => _getCredential != null;
+
+        public CredentialProvider(
+            IReadOnlyDictionary<string, HttpConfig> configCredentials, Func<string, Task<Dictionary<string, HttpConfig>>>? getCredential = null)
         {
-            _credentials = credentials;
+            _configCredentials = configCredentials;
+            _getCredential = getCredential;
         }
 
-        public Dictionary<string, string> GetCredentials(HttpRequestMessage request)
+        public bool FillInCredentials(HttpRequestMessage request)
```

Comment:
>FillInCredentials [](start = 20, length = 17)

I am thinking keeping original output `Dictionary<string, string>` might be cleaner...